staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -761,26 +792,86 @@ func (s *Controller) nodeSyncInternal(ctx context.Context, workers int) {
 		numServices-len(s.servicesToUpdate), numServices)
 }
 
-// nodeSyncService syncs the nodes for one load balancer type service
-func (s *Controller) nodeSyncService(svc *v1.Service) bool {
+// nodeSyncService syncs the nodes for one load balancer type service. It assumes the
+// caller always provides the set of Ready and NotReady nodes as oldNodes and newNodes.
+// The return value indicates if we should retry. Hence, this functions returns false
+// if we've updated load balancers and finished doing it successfully, or didn't try to
+// at all because there's no need. This function returns true if we tried to update load
+// balancers and failed, indicating to the caller that we should try again.
+func (s *Controller) nodeSyncService(svc *v1.Service, oldNodes, newNodes []*v1.Node) bool {
 	if svc == nil || !wantsLoadBalancer(svc) {
 		return false
 	}
 	klog.V(4).Infof("nodeSyncService started for service %s/%s", svc.Namespace, svc.Name)
-	hosts, err := listWithPredicate(s.nodeLister, s.getNodeConditionPredicate())
-	if err != nil {
-		runtime.HandleError(fmt.Errorf("failed to retrieve node list: %v", err))
-		return true
+	newReadyNodes := filterWithPredicate(newNodes, s.getNodeConditionPredicate(includeOnlyReadyNodes))
+	// For services of ETP=local: we do not want to sync nodes transitioning from NotReady <-> Ready
+	// if these hosts are not actually hosting any endpoints for the service. Doing that leads to
+	// unnecessary updates for LBs which will never send traffic to the nodes in question in any case.
+	if svc.Spec.ExternalTrafficPolicy == v1.ServiceExternalTrafficPolicyTypeLocal {
+		needsSync := func() bool {
+			// We have four cases to consider here:
+			//   a) Node ADD
+			//   b) Node DEL
+			//   c) Node UPDATE Ready -> NotReady
+			//   d) Node UPDATE NotReady -> Ready
+			// In case of (c) or (d) we can skip sync for ETP=Local Service
+			oldNodeNames := nodeNames(oldNodes)
+			newNodeNames := nodeNames(newNodes)
+			if !oldNodeNames.Equal(newNodeNames) {
+				// The set of nodes has changed - needs sync.
+				return true
+			}
+			// The set of nodes didn't change, but maybe readiness did.
+			oldReadyNodes := filterWithPredicate(oldNodes, s.getNodeConditionPredicate(includeOnlyReadyNodes))
+			oldReadyNodeNames := nodeNames(oldReadyNodes)
+			newReadyNodeNames := nodeNames(newReadyNodes)
+
+			// Figure out which nodes changed readiness.
+			changedNodeNames := oldReadyNodeNames.Difference(newReadyNodeNames)
+			changedNodeNames = changedNodeNames.Union(newReadyNodeNames.Difference(oldReadyNodeNames))
+			if len(changedNodeNames) == 0 {
+				// Something else changed, but LBs are not concerned by the
+				// change, skip sync.
+				return false
```

Comment:
I've re-worked the code quite a bit and added the part which performs this validation in its own separate (last) commit. It might need some re-working, but I don't have any good ideas for the time being.